typeck: fix `?` suggestion span

jieyouxu opened the pull request:
Noticed in <https://github.com/rust-lang/rust/pull/112043#issuecomment-2043565292>, if the

```
use the `?` operator to extract the `Result<(), std::fmt::Error>` value, propagating a `Result::Err` value to the caller
```

suggestion is applied to a macro that comes from a non-local crate (e.g. the stdlib), the suggestion span can become non-local, which will cause newer rustfix versions to fail.

This PR tries to remedy the problem by recursively probing ancestors of the expression span, trying to identify the most ancestor span that is (1) still local, and (2) still shares the same syntax context as the expression.

This is the same strategy used in https://github.com/rust-lang/rust/pull/112043.

The test unfortunately cannot `//@ run-rustfix` because there are two conflicting MaybeIncorrect suggestions that when collectively applied, cause the fixed source file to become non-compilable.

Also avoid running `//@ run-rustfix` for `tests/ui/typeck/issue-112007-leaked-writeln-macro-internals.rs` because that also contains conflicting suggestions.

cc @ehuss who noticed this. This question mark span fix + not running rustfix on the tests containing conflicting MaybeIncorrect suggestions should hopefully unblock rustfix from updating.

jieyouxu committed:
test: check that `?` suggestion has local span

This can cause rustfix to crash because the `?` suggestion previously
could point into non-local spans, such as into the stdlib.

jieyouxu committed:
typeck: fix `?` operator suggestion span

compiler-errors commented on compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:
```diff
@@ -3180,3 +3171,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+/// syntax context of a potential parent callsite changes, such as if the potential parent callsite
+/// is in a foreign macro. This helps to prevent leaking implementation details from upstream
+/// crates and stdlib crates that the user likely have no control over.
+fn find_local_most_ancestor_suggestable_span(initial_span: Span) -> Span {
```
Why is this different than [`Span::find_ancestor_in_same_ctxt`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/span_encoding/struct.Span.html#method.find_ancestor_in_same_ctxt)?

compiler-errors replied:
Also https://github.com/rust-lang/rust/pull/112043/files#r1556509496

compiler-errors replied:
Oh, I see. We find the *last* ancestor that shares the ctxt, not the first. Huh.

jieyouxu replied:
Yeah, it was to get better error reporting for when there are multiple local nested macros.

cjgillot replied:
This method should go in `rustc_span`, as a method on `Span`.

jieyouxu replied:
Do you have a better name for this? `find_local_most_ancestor_suggestable_span` is what I came up with but it feels not very accurate

Nadrieril replied:
How about `find_oldest_ancestor_in_same_ctxt`?

jieyouxu replied:
Sounds better than what I had 😆

jieyouxu committed:
test: avoid running rustfix on leaked writeln macro internals test

This test contains conflicting MaybeIncorrect suggestions which will
cause the fixed file to not compile.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8608461134/job/23590868207) [(plain)](https://github.com/rust-lang/rust/commit/98e920109f2b1cefe600581feb6989e10ab561cc/checks/23590868207/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 31.7s done
#16 DONE 36.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.27s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- compiler/rustc_span/src/lib.rs - Span::find_local_most_ancestor_suggestable_span (line 752) stdout ----
error[E0308]: mismatched types
##[error]  --> compiler/rustc_span/src/lib.rs:767:10
   |
17 |      Err(outer!(x))
   |          ^^^^^^^^^ expected `Box<dyn Error>`, found `String`
   = note: expected struct `Box<dyn std::error::Error>`
              found struct `String`
   = note: this error originates in the macro `format` which comes from the expansion of the macro `outer` (in Nightly builds, run with -Z macro-backtrace for more info)
   = note: this error originates in the macro `format` which comes from the expansion of the macro `outer` (in Nightly builds, run with -Z macro-backtrace for more info)
help: call `Into::into` on this expression to convert `String` into `Box<dyn std::error::Error>`
   |
11 |          format!("error: {}", $x).into()

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
For more information about this error, try `rustc --explain E0308`.
Couldn't compile the test.

failures:
    compiler/rustc_span/src/lib.rs - Span::find_local_most_ancestor_suggestable_span (line 752)

test result: FAILED. 2 passed; 1 failed; 3 ignored; 0 measured; 0 filtered out; finished in 89.65ms

error: doctest failed, to rerun pass `-p rustc_span --doc`
  local time: Tue Apr  9 01:02:39 UTC 2024
  network time: Tue, 09 Apr 2024 01:02:39 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jieyouxu commented:
@rustbot author

jieyouxu commented:
(Ignored the doctest because it contains type errors for illustration purposes)
@rustbot ready

jieyouxu commented:
Changed the method name to `find_oldest_ancestor_in_same_ctxt` and fixed its docs.

Nadrieril commented:
Looks good!

@bors r+

bors commented:
:pushpin: Commit 420e3f1d5cf8e156d6f4ae636ce5bbaa6b7c9c70 has been approved by `Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nadrieril 420e3f1d5cf8e156d6f4ae636ce5bbaa6b7c9c70 -->
<!-- homu: {"type":"Approved","sha":"420e3f1d5cf8e156d6f4ae636ce5bbaa6b7c9c70","approver":"Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
